### PR TITLE
Add JCache tests for dynamic changes and server restart

### LINKED ISSUE #19453: SEC 380: Test dynamic configuration changes for JCache
Test making dynamic configuration changes to jCachingProvider, jCacheManager and jCache.

Verify that the changes are propagated to the underlying services... i.e. that changes to jCacheProvider effect jCacheManager and jCache; changes to jCacheManager effect jCache where appropriate.

### LINKED ISSUE #19454: SEC 380: Test JCache startup and shutdown
Verify that JCaches maintain state across server startup and shutdown. Servers should be able to connect and disconnect to the cache without effecting the contents.

There should be tests for the authentication cache, the logged out cookie cache and the HTTP session cache.

### PR DESCRIPTION
**Pull request by ryanesch**
fixes #19453
fixes #19454

Adding tests for:

- Server restart in auth cache
- Server restart in logged out cookie cache
- Dynamic changes to server.xml in libraryRef and properties

### CODE REVIEW

File: dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.ltpa.cookie.cache.restart.2/bootstrap.properties
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.ltpa.cookie.cache.restart.2/bootstrap.properties]: I added `com.ibm.websphere.security.*=all` to all the other servers. Should add it here as well.

```suggestion
com.ibm.ws.logging.trace.specification=*=info:jcache=all:com.ibm.ws.security.*=all:com.ibm.websphere.security.*=all:com.ibm.ws.webcontainer.security.*=all
```

File: dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.ltpa.cookie.cache.restart.1/bootstrap.properties
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.ltpa.cookie.cache.restart.1/bootstrap.properties]: I added `com.ibm.websphere.security.*=all` to all the other servers. Should add it here as well.

```suggestion
com.ibm.ws.logging.trace.specification=*=info:jcache=all:com.ibm.ws.security.*=all:com.ibm.websphere.security.*=all:com.ibm.ws.webcontainer.security.*=all
```

File: dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.dynamicupdate.1/bootstrap.properties
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.dynamicupdate.1/bootstrap.properties]: I added `com.ibm.websphere.security.*=all` to all the other servers. Should add it here as well.

```suggestion
com.ibm.ws.logging.trace.specification=*=info:jcache=all:com.ibm.ws.security.*=all:com.ibm.websphere.security.*=all:com.ibm.ws.webcontainer.security.*=all
```

File: dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.auth.cache.restart.1/bootstrap.properties
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.auth.cache.restart.1/bootstrap.properties]: I added `com.ibm.websphere.security.*=all` to all the other servers. Should add it here as well.

```suggestion
com.ibm.ws.logging.trace.specification=*=info:jcache=all:com.ibm.ws.security.*=all:com.ibm.websphere.security.*=all:com.ibm.ws.webcontainer.security.*=all
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java:40]: Probably should skip EE9 repeat. I don't think it buys us anything.
```suggestion
@SkipForRepeat(SkipForRepeat.EE9_FEATURES)
public class JCacheAuthenticationCacheServerRestartTest extends BaseTestCase {
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java:44]: Probably should skip EE9 repeat. I don't think it buys us anything.

```suggestion
@SkipForRepeat(SkipForRepeat.EE9_FEATURES)
public class JCacheDynamicUpdateTest extends BaseTestCase {
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheLtpaLoggedOutCookieCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheLtpaLoggedOutCookieCacheServerRestartTest.java:42]: Probably should skip EE9 repeat. I don't think it buys us anything.

```suggestion
@SkipForRepeat(SkipForRepeat.EE9_FEATURES)
public class JCacheLtpaLoggedOutCookieCacheServerRestartTest extends BaseTestCase {
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java]: Maybe update the comments to say that this will show that the distributed authentication cache has maintained entries across the sever restart.

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java:57]: This is going in for another PR, but please add this after my PR is merged.

```suggestion
        waitForDefaultHttpsEndpoint(server1);
        waitForCachingProvider(server1, AUTH_CACHE_NAME);
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java]: This is going in for another PR, but please add this after my PR is merged.

```suggestion
        waitForDefaultHttpsEndpoint(server1);
        waitForCachingProvider(server1, AUTH_CACHE_NAME);
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheLtpaLoggedOutCookieCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheLtpaLoggedOutCookieCacheServerRestartTest.java:57]: This is going in for another PR, but please add this after my PR is merged.

```suggestion
        waitForDefaultHttpsEndpoint(server1);
        waitForCachingProvider(server1, COOKIE_CACHE_NAME);
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheLtpaLoggedOutCookieCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheLtpaLoggedOutCookieCacheServerRestartTest.java]: This is going in for another PR, but please add this after my PR is merged.

```suggestion
        waitForDefaultHttpsEndpoint(server2);
        waitForCachingProvider(server2, COOKIE_CACHE_NAME);
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java]: don't need the try/finally with only 1 server.

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java]: In the same PR that I added the wait for the HTTPS endpoing, I have renamed this method. It is now called `assertJCacheLtpaAuthCacheHit` and I added another method called `assertInMemoryLtpaAuthCacheHit`.

Please wait until my PR is merged and then use `assertJCacheLtpaAuthCacheHit`.

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java:105]: I would probably do another call here, like you do in step 3, to verify that the entry is cached before we restart the server.

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java]: BaseTestCase has a method called resetMarksInLogs. I suggest we call it here.

```suggestion
        resetMarksInLogs(server);
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java]: We can default size and TTL.
```suggestion
        startServer1(server1, groupName, null, null);
```

**Review by jvanhill — CHANGES_REQUESTED** [review]
Please see my comments in this review and make the changes we discussed on our call.

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java]: Don't need the finally when not using two servers.

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java:214]: We don't check bad cache->cacheManagerRef and cacheManager->cachingProviderRef . I am guessing that Liberty will complain about this before it makes it to us, but we should probably check it.

I am also wondering why you didn't just use the fattest.simplicity configuration elements instead of using multiple configuration files? It might have been easier, and more clear what we were changing for each step of the test.

**Review by jvanhill — CHANGES_REQUESTED** [review]
I don't really have any major findings, just some suggestions that you can either do or ignore.

However; you do need to rebase on release, as there are going to be some breaking changes in there. I have pointed out some of them in my comments from the last time I reviewed.

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java
Comment (ryanesch) [reply dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheDynamicUpdateTest.java:214]: I added the server config elements to fattest.simplicity and reworked the test.

File: dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/AuthCache.java
Comment (jvanhill) [inline dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/AuthCache.java:38]: Probably want the getCacheRef method as well.

File: dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/Cache.java
Comment (jvanhill) [inline dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/Cache.java:40]: Get methods?

File: dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/CacheManager.java
Comment (jvanhill) [inline dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/CacheManager.java:49]: Get methods?

File: dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/CachingProvider.java
Comment (jvanhill) [inline dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/CachingProvider.java]: Get methods?

File: dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/Properties.java
Comment (jvanhill) [inline dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/Properties.java]: These are infinispan specific. The Properties element is used all over (not just for JCache, sessionCache uses it, as do others).

Instead you should use @XmlAnyAttribute and see if you can make that work. I think you annotate a map and it will allow you to assign any attribute name any value.

File: dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/Properties.java
Comment (jvanhill) [inline dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/cache/Properties.java]: This would need to move somewhere else to keep this class generic.

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java]: Delete this, as we don't need TTL testing here (and I have disabled elsewhere b/c it really just tests the provider and not our code.
```suggestion
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java]: No TTL testing here.
```suggestion
        startServer1(server1, groupName, 25000, null);
```

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java
Comment (jvanhill) [reply dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheAuthenticationCacheServerRestartTest.java:57]: Doesn't appear the `waitForDefaultHttpsEndpoint(server1);` call has been added.

File: dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheLtpaLoggedOutCookieCacheServerRestartTest.java
Comment (jvanhill) [reply dev/io.openliberty.jcache.internal_fat.hazelcast/fat/src/io/openliberty/jcache/internal/fat/JCacheLtpaLoggedOutCookieCacheServerRestartTest.java:57]: The `waitForDefaultHttpsEndpoint(server1);` method call is not in here yet.

File: dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.dynamicupdate.1/server.xml
Comment (jvanhill) [inline dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.dynamicupdate.1/server.xml]: So... you might want to merge with release. I updated all the server.xml files so that common configuration was in include files. You will want to do the same here. The timedexit and dist... features have been moved into testingConfig.xml.

The httpEndpoint has been moved into httpEndpoint1.xml (and httpEndpoint2.xml if you use a 2nd server).

The applications have been moved into app*.xml (in this case I think appBasicAuth1.xml).

Comment (jvanhill) [reply dev/io.openliberty.jcache.internal_fat.infinispan/publish/servers/io.openliberty.jcache.internal.fat.dynamicupdate.1/server.xml]: This applies to the other server.xml files as well. I won't comment on them all.

**Review by jvanhill — CHANGES_REQUESTED** [review]
There are a few changes I have requested.

Another test that you might consider adding to the JCacheDynamicUpdateTest is starting from a good configuration and creating new `cache` elements to see that we will create them dynamically. Also remove them and see that we disconnect. Not sure if we have a message for that?

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Review by jvanhill — APPROVED** [review]

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#run-libby-bot

**Pull request merged**
